enhancements/sig-policy/85-gatekeeper-policy-integration/README.md

```diff
@@ -0,0 +1,258 @@
+          match:
+            kinds:
+              - apiGroups: [""]
+                kinds: ["Namespace"]
+          parameters:
+            labels: ["gatekeeper"]
+```
+
+When the `Policy` has the `spec.remediationAction` field set to `inform`, it would override the constraint's
+`spec.enforcementAction` with `warn`. This will not deny requests but instead, it will
+[warn](https://open-policy-agent.github.io/gatekeeper/website/docs/violations#warn-enforcement-action) the user making
+the request if their request violated the constraint. Existing and new violations would show up in the constraint's
+`status.violations` section.
+
+If the `remediationAction` field is set to `enforce`, the constraint's `spec.enforcementAction` would be overriden with
+`deny`. Existing violations would show up in the constraint's `status.violations` section.
+
+If the `remediationAction` field is not set, the constraint's `spec.enforcementAction` would not be modified.
+
+#### Status Reporting
+
+After applying a Gatekeeper constraint, Kubernetes objects on the cluster that violate the constraint will show up in
+the constraint's `status.violations` field as shown in the example below.
+
+```yaml
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sRequiredLabels
+metadata:
+  name: ns-must-have-gk
+spec:
+  enforcementAction: dryrun
+  match:
+    kinds:
+      - apiGroups: [""]
+        kinds: ["Namespace"]
+  parameters:
+    labels: ["gatekeeper"]
+status:
+  auditTimestamp: "2019-08-15T01:46:13Z"
+  enforced: true
+  violations:
+    - enforcementAction: dryrun
+      kind: Namespace
+      message: 'you must provide labels: {"gatekeeper"}'
+      name: default
+    - enforcementAction: dryrun
+      kind: Namespace
+      message: 'you must provide labels: {"gatekeeper"}'
+      name: gatekeeper-system
+```
+
+The status message for the constraint in the OCM `Policy` would be in the format of
+`<enforcementAction> - <message> (on <kind> <namespace/name or name>)` for each violation entry each separated by a
+semicolon. For example, the above example would produce a status message of
+`dryrun - you must provide labels: {"gatekeeper"} (on Namespace default); dryrun - you must provide labels: {"gatekeeper"} (on Namespace gatekeeper-system)`.
+
+If Gatekeeper is not installed, the OCM `Policy` will be marked as non-compliant with a message stating that the user
+needs to install Gatekeeper to apply such a constraint.
+
+#### Policy Deletion
+
+When an OCM `Policy` embeds a Gatekeeper `ConstraintTemplate`/constraint is deleted, the embeded Gatekeeper
+`ConstraintTemplate`/constraint will be deleted to match the behavior of other OCM policy templates.
+
+### User Stories
+
+#### Story 1
+
+As a policy user, I would like to have native multicluster support for an admission controller so that I can have a
+consistent distribution and status aggregation mechanism for all my policies.
+
+### Implementation Details/Notes/Constraints [optional]
```

Comment:
If I understand correctly, template-sync will create the ConstraintTemplates and constraints (so, it still creates all the templates in a Policy). And the new controller will just be responsible for watching those and emitting compliance events. The status-sync is still the only thing that puts those events into the status of the policy.

If that's right, then is it necessary for the new controller be included in the framework container/pod? It feels more similar to the other policy controllers, which watch their policy types and emit events. I'm imagining that someone might want to turn this controller off, in which case it being a separate pod might make that easier.

Granted, this new controller (as described here) operates a little differently than other policy controllers: it needs to watch Policies to know what types the constraints will be, and it uses the new `policy.open-cluster-management.io/policy` label instead of ownerreferences to track some things. But I'm not sure if that necessarily means it must be included in the framework container.

Reply:
It doesn't have to be in the Policy Framework. I was doing this for convenience and also because it's essentially just relaying status updates, so it seemed like a companion to the `policy-status-sync` controller.

With the current design, the resource utilization should be very minimal if the Gatekeeper integration is not used since all it will do is loop through the `policy-templates` array of all the policies looking for Gatekeeper constraints, but the cache will be shared with the existing controllers in the Policy Framework, so another watch won't be necessary.

I think the low impact is worth making the installation experience easier, but I'm open to suggestions on this one.

Reply:
I'm ok with it either way, I wanted to check if there was a need to keep it in the framework.

I have a slight preference for it to be a separate controller. That follows the pattern of the other policy controllers, and would make it easier to debug and potentially disable. I think that when it is being used, it will have a non-zero impact on the performance of the rest of the framework, and in my opinion the rest of the framework is more essential and should be "protected"

But since it does need to watch all the Policies, a shared cache is better there...

Reply:
Note that we don't want to install any additional standalone controllers on managed clusters in any "default" installation scenarios.  We need to be very mindful of any impacts to the footprint we have on the managed cluster.  That's one reason I'm for consolidating cert and iam into a security controller and maybe this would fit better into that category.  Since iam and cert are lightly used, adding admissions control would be a third "low impact" security controller that could be combined.